File: code/ch09/9.2.2.3.long_substr_without_repeat_char.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
            int res = 0, left = 0, right = 0;
            unordered_set<char> setting;

            while (right < s.size()) {
                if (setting.find(s[right]) != setting.end()) {
                    while(left < right) {
                        if (s[left] == s[right]) {
                            setting.erase(s[left]);
                            left++;
                            break;
                        } else {
                            setting.erase(s[left]);
                            left++;
                        }
                    }
                }
                setting.insert(s[right]);
                res = max(res, right - left + 1);
                right++;
            }
            return res;
    }
};

File: code/ch09/9.2.2.3.long_substr_without_repeat_char.py
class Solution:
    def lengthOflongestSubstring(self, s:str)->int:
        res, left, right = 0, 0, 0
        settings = set()

        while right < len(s):
            if s[right] in settings:
                while left < right:
                    if s[left] == s[right]:
                        settings.discar(s[left])
                        left = left + 1
                        break
                    else:
                        settings.discard(s[left])
                        left = left + 1

            settings.add(s[right])
            res = max(res, right - left + 1)
            right = right + 1

        return res
